### PR TITLE
Use Arrow Tensors for serializing numpy arrays and get rid of extra memcpy.

### PR DESCRIPTION
**Pull request by pcmoritz**
This PR also compiles both plasma and arrow with -march=native, -mtune=native and -O3.

### CODE REVIEW

**Comment by wesm** [comment]
Sweet

File: src/numbuf/python/src/pynumbuf/numbuf.cc [hunk 1]
```diff
@@ -275,12 +345,9 @@ static PyObject* store_list(PyObject* self, PyObject* args) {
   }
   CHECK(error_code == PlasmaError_OK);
 
-  auto target = std::make_shared<FixedBufferStream>(LENGTH_PREFIX_SIZE + data, size);
-  std::shared_ptr<arrow::ipc::FileWriter> writer;
-  ipc::FileWriter::Open(target.get(), batch->schema(), &writer);
-  writer->WriteRecordBatch(*batch, true);
-  writer->Close();
-  *((int64_t*)data) = size;
+  auto target = std::make_shared<FixedBufferStream>(LENGTH_PREFIX_SIZE + data, total_size);
```
Comment (wesm) [inline src/numbuf/python/src/pynumbuf/numbuf.cc]: I added a variant of this class (hopefully suitable for your needs) recently to Arrow: https://github.com/apache/arrow/blob/master/cpp/src/arrow/io/memory.h#L72

Comment (pcmoritz) [inline src/numbuf/python/src/pynumbuf/numbuf.cc]: @wesm Thanks! I think we will need to modify the class so we can use our custom multithreaded memcpy. We could provide an interface similar to https://github.com/ray-project/ray/blob/master/src/numbuf/python/src/pynumbuf/memory.h#L19 and have a default one that just calls std::memcpy. What do you think about that?

Comment (wesm) [inline src/numbuf/python/src/pynumbuf/numbuf.cc]: Yes, my thinking was that you could contribute the multithreaded memcpy to Arrow and we could add an option to this class to switch between that and std::memcpy. Let me know if I can help

Comment (pcmoritz) [inline src/numbuf/python/src/pynumbuf/numbuf.cc]: Sounds like a good plan! I won't have time to do this today but will create a JIRA and PR in the next couple of days.

File: src/numbuf/thirdparty/download_thirdparty.sh [hunk 2]
```diff
@@ -8,7 +8,7 @@ set -e
 TP_DIR=$(cd "$(dirname "${BASH_SOURCE:-$0}")"; pwd)
 
 if [ ! -d $TP_DIR/arrow ]; then
-  git clone https://github.com/apache/arrow/ "$TP_DIR/arrow"
+  git clone https://github.com/pcmoritz/arrow/ "$TP_DIR/arrow"
```
Comment (robertnishihara) [inline src/numbuf/thirdparty/download_thirdparty.sh]: Switch back to `apache/arrow`.

File: src/numbuf/cpp/src/numbuf/sequence.cc [hunk 3]
```diff
@@ -15,27 +15,18 @@ SequenceBuilder::SequenceBuilder(MemoryPool* pool)
...
-      int8_tensors_(std::make_shared<Int8Type>(), pool),
-      uint16_tensors_(std::make_shared<UInt16Type>(), pool),
-      int16_tensors_(std::make_shared<Int16Type>(), pool),
-      uint32_tensors_(std::make_shared<UInt32Type>(), pool),
-      int32_tensors_(std::make_shared<Int32Type>(), pool),
-      uint64_tensors_(std::make_shared<UInt64Type>(), pool),
-      int64_tensors_(std::make_shared<Int64Type>(), pool),
-      float_tensors_(std::make_shared<FloatType>(), pool),
-      double_tensors_(std::make_shared<DoubleType>(), pool),
+      tensor_indices_(pool, std::make_shared<Int32Type>()),
```
Comment (pcmoritz) [inline src/numbuf/cpp/src/numbuf/sequence.cc]: Done in the header

**Pull request merged**